### PR TITLE
UPM compliant packages for MRTK

### PR DESCRIPTION
**Pull request by ei2kpi-ptc**
## Overview
Creates UPM compliant packages for Unity 2018.3

## Changes
- Fixes: #4403 .

## Verification
- Please specifically check out the changes in `MixedRealityToolkitFiles.cs` since it looks like there is a fair amount of code there that does string manipulation to find the MRTK folders and assets. I am pretty sure I fixed most of code paths but would like to get some extra coverage here.
- I already have an Azure Pipeline build that zips up and creates the UPM compliant packages in my fork. See release here: https://github.com/ei2kpi-ptc/MixedRealityToolkit-Unity/releases/tag/v2.0.0-RC1-Refresh-UPM. It would be great if we could add this to the MRTK releases as well since it is just a simple archive of the folders.

### CODE REVIEW

**Comment by wiwei** [comment]
@paco-ms @andreiborodin this change establishes some naming for the MRTK package (i.e. com.microsoft.*), would be good to have your eyes on this as this is something that I think both of you guys have been involved in.

File: Assets/MixedRealityToolkit/package.json [hunk 1]
```diff
@@ -0,0 +1,17 @@
+{
+	"name": "com.microsoft.mixedrealitytoolkit",
```
Comment (andreiborodin) [inline Assets/MixedRealityToolkit/package.json]: >com.microsoft.mixedrealitytoolkit [](start = 10, length = 33)

If we are going with this name pattern
, we should name this "com.microsoft.mixedreality.toolkit"

File: Assets/MixedRealityToolkit/Utilities/Editor/Setup/MixedRealityToolkitFiles.cs [hunk 2]
```diff
@@ -76,8 +78,16 @@ public static bool AreFoldersAvailable
...
+            if (AssetDatabase.IsValidFolder($"Packages/{MRTK_PACKAGE_ID}"))
+            {
+                mrtkFolders.Add(Path.GetFullPath($"Packages/{MRTK_PACKAGE_ID}"));
+                searchForFoldersTask = Task.CompletedTask;
+            }
+            else
+            {
+                string path = Application.dataPath;
+                searchForFoldersTask = Task.Run(() => SearchForFoldersAsync(path));
+            }
```
Comment (andreiborodin) [inline Assets/MixedRealityToolkit/Utilities/Editor/Setup/MixedRealityToolkitFiles.cs]: The change I made below was to account for potentially split folders, but I guess that may have been excessive. This seems reasonable.

**Comment by andreiborodin** [comment]
Looks good to me, I commented on naming, it prob should be com.microsoft.mixedreality.toolkit. Otherwise, are we ready to publish this like this?

---
In reply to: [494129618](https://github.com/microsoft/MixedRealityToolkit-Unity/pull/4404#issuecomment-494129618) [](ancestors = 494129618)

**Comment by andreiborodin** [comment]
Will, we should probably have just one package here as well.

---
In reply to: [494167625](https://github.com/microsoft/MixedRealityToolkit-Unity/pull/4404#issuecomment-494167625) [](ancestors = 494167625,494129618)

File: Assets/MixedRealityToolkit/Utilities/Editor/Setup/MixedRealityToolkitFiles.cs (hunk 2 above)
Comment (ei2kpi-ptc) [reply Assets/MixedRealityToolkit/Utilities/Editor/Setup/MixedRealityToolkitFiles.cs]: Is this resolved? Or do I need to change this to have everything in one UPM package and then go into the individual folders? I tried to match the Nuget Package model that seems to be in the repo right now

File: Assets/MixedRealityToolkit/package.json (hunk 1 above)
Comment (ei2kpi-ptc) [reply Assets/MixedRealityToolkit/package.json]: I'll change this and update the PR when I get a chance

File: Assets/MixedRealityToolkit/Utilities/Editor/Setup/MixedRealityToolkitFiles.cs (hunk 2 above)
Comment (andreiborodin) [reply Assets/MixedRealityToolkit/Utilities/Editor/Setup/MixedRealityToolkitFiles.cs]: Yes, it is, sorry!

**Comment by yacuzo** [comment]
I adapted MRTK for UPM for our own projects. You also have to do some changes in Assets/Export/MixedRealityToolkit/Inspectors/ControllerPopupWindow.cs. It uses a lot of Application.dataPath.

Maybe open the repo in VS Code and do a full-repo text search for Application.dataPath. Could be more hidden somewhere.

**Comment by julenka** [comment]
> I adapted MRTK for UPM for our own projects. You also have to do some changes in Assets/Export/MixedRealityToolkit/Inspectors/ControllerPopupWindow.cs. It uses a lot of Application.dataPath.
> 
> Maybe open the repo in VS Code and do a full-repo text search for Application.dataPath. Could be more hidden somewhere.

Hi @ei2kpi-ptc could you please take a look at yacuzo's comments? Looks like there's a few more cases to clean up.

**Comment by julenka** [comment]
Hi @ei2kpi-ptc could you please resolve conflicts?

File: Assets/MixedRealityToolkit/package.json (hunk 1 above)
Comment (ei2kpi-ptc) [reply Assets/MixedRealityToolkit/package.json]: Package.json files updated

File: Assets/MixedRealityToolkit/Utilities/Editor/Setup/MixedRealityToolkitFiles.cs (hunk 2 above)
Comment (ei2kpi-ptc) [reply Assets/MixedRealityToolkit/Utilities/Editor/Setup/MixedRealityToolkitFiles.cs]: Looks like a bunch more stuff was added to mixedrealitytoolkitfiles to support Core and other folders. I merged my change with this and am testing it right now

File: azure-pipelines.yml [hunk 3]
```diff
@@ -0,0 +1,17 @@
+# Starter pipeline
```
Comment (julenka) [inline azure-pipelines.yml]: Hi ei2kpi, it looks like this is specific to your pipeline, not the mrtk pipeline. Could you please remove it? @MenelvagorMilsom please correct me if I'm wrong here.

**Comment by julenka** [comment]
Hm there are 191 files changed in this PR, including .meta files and .unity files. Are those intended to be part of the PR? It looks like some unintentional changes have gone in here.

File: azure-pipelines.yml (hunk 3 above)
Comment (MenelvagorMilsom) [reply azure-pipelines.yml]: Yeah, this isn't part of the mrtk pipelines and can not run on our builds. Possibly an unintentional addition?

**Comment by ei2kpi-ptc** [comment]
Argh! looks like I messed up something in my fork and it added it to the PR automatically. 
I had originally made my changes in mrtk_release but it looks like PRs to MRTK have to go into mrtk_development. so, it looks like I rebased my fork's mrtk_release onto the upstream mrtk_development which then added all those commits onto the PR as well! :(

**Comment by ei2kpi-ptc** [comment]
Is it OK if I abandon this PR and create a new one off of mrtk_development with only the correct changes?

**Comment by julenka** [comment]
> Is it OK if I abandon this PR and create a new one off of mrtk_development with only the correct changes?

Yes go ahead!

**Comment by ei2kpi-ptc** [comment]
Abandoning this PR and will create a new one with just the necessary changes from mrtk_development

**Pull request closed without merging**